dbt/include/snowflake/macros/adapters.sql

```diff
@@ -47,23 +47,108 @@
+        many schemas are contained in their databases. Since there was already
+        one iteration before attempting pagination, loop.index == max_iter means
+        the limit has been surpassed.
+      #}
+
+      {%- if loop.index == max_iter -%}
+        {%- set msg -%}
+           dbt will list a maximum of {{ max_total_results }} schemas in database {{ database }}.
+           Your database exceeds this limit. Please contact support@getdbt.com for troubleshooting tips,
+           or review and reduce the number of objects contained.
+        {%- endset -%}
+
+        {% do exceptions.raise_compiler_error(msg) %}
+      {%- endif -%}
+
+      {%- do paginated_schemas.append(paginated_result) -%}
+      {% set watermark.schema_name = paginated_result.columns[1].values()[-1] %}
+
+      {#
+        terminating condition: paginated_n < max_results_per_iter means we reached the end
+      #}
+      {%- if paginated_n < max_results_per_iter -%}
+         {%- break -%}
+      {%- endif -%}
+    {%- endfor -%}
+
+  {{ return(paginated_schemas) }}
+
+{% endmacro %}
+
+{% macro snowflake__list_schemas(database, max_iter=10, max_results_per_iter=1000) %}
+
+  {%- set max_total_results = max_results_per_iter * max_iter -%}
+
+  {%- set sql -%}
+    show terse schemas in database {{ database }} limit {{ max_results_per_iter }};
+  {%- endset -%}
+
+  {%- set result = run_query(sql) -%}
+
+  {%- set n = (result | length) -%}
+  {%- set watermark = namespace(schema_name=result.columns[1].values()[-1]) -%}
+  {%- set paginated = namespace(result=[]) -%}
+
+  {% if n >= max_results_per_iter %}
+
+    {% set paginated.result = snowflake__get_paginated_schemas_array(
+         max_iter,
+         max_results_per_iter,
+         max_total_results,
+         database,
+         watermark
+       )
+    %}
+
+  {% endif %}
+
+  {%- set all_results_array = [result] + paginated.result -%}
+  {%- set result = result.merge(all_results_array) -%}
+  {%- do return(result) -%}
+
+{% endmacro %}
+
+{# macro snowflake__list_schemas(database) -#}
   {# 10k limit from here: https://docs.snowflake.net/manuals/sql-reference/sql/show-schemas.html#usage-notes #}
-  {% set maximum = 10000 %}
-  {% set sql -%}
+  {# set maximum = 10000 #}
+  {# set sql -#}
     show terse schemas in database {{ database }}
     limit {{ maximum }}
-  {%- endset %}
-  {% set result = run_query(sql) %}
-  {% if (result | length) >= maximum %}
-    {% set msg %}
+  {#- endset %}
+  {# set result = run_query(sql) #}
+  {# if (result | length) >= maximum #}
+    {# set msg %}
       Too many schemas in database {{ database }}! dbt can only get
       information about databases with fewer than {{ maximum }} schemas.
-    {% endset %}
-    {% do exceptions.raise_compiler_error(msg) %}
-  {% endif %}
+    {# endset %}
+    {# do exceptions.raise_compiler_error(msg) #}
+  {# endif #}
   {{ return(result) }}
-{% endmacro %}
+{# endmacro #}
```

Comment:
@matt-winkler is all of this supposed to be commented out?

Reply:
@McKnight-42 good catch. I removed this block of code just now.